### ISSUE TITLE
[Bug] Explicit layout with record struct

### ISSUE THREAD

**Issue by Xyncgas** [fsharp/fslang-suggestions#1240]
### Discussed in https://github.com/fsharp/fslang-suggestions/discussions/1233

<div type='discussions-op-text'>

<sup>Originally posted by **Xyncgas** January 12, 2023</sup>
For example
```F#
[<Struct; StructLayout(LayoutKind.Explicit)>]
 type Dimension2D = 
            {
                [<FieldOffsetAttribute(0)>]
                Memory : Vector2
                [<FieldOffsetAttribute(0)>]
                Width : float32
                [<FieldOffsetAttribute(4)>]
                Height : float32
            }
            with
            static member Zero = { Memory = Vector2.Zero; Width = 0f; Height = 0f; }
```
The point of explicit layout is you can change the memory by changing width / height, or the opposite way by using System.Numerics.Vector2 methods to change the memory

However I noticed that, while it's already hard to create a record with only memory is set
```F#
{Dimension2D.Memory = Vector2.Zero} //Alone doesn't work
```
If I were to do :
```F#
{ Dimension2D.Zero with Memory = Vector2.Multiply(aVector, value) } //Bug : the value remains zero after this no matter what you do
```
Where ```AnotherMemory``` is not zero and so is value, the result is zero for some reason

But if I were to do
```F#
let v = Vector2.Multiply(aVector, value)
{ Dimension2D.Zero with Memory = v; Width = aVector.X; Height = aVector.Y } //Works
```

It gives me a feeling that either I can't have record struct with explicit layout in F# like I can in C#, or there is a bug with it

I wanted to do this because if I store my value in Vector struct they get hardware accelerated sometimes. Performance is important to F# aye, so I am hoping I can write these kinds of codes</div>

**Comment by T-Gro** [comment]
The "with" clause copies over all values, the ones not specified are copied with value of the source.
That means each value is being set, and in the case of overlapping explicit fields, order matters. So in your penultimate case, the memory was set, but after that the implicit values for Width and Height were copied over as well and therefore overwrote the value from Memory.

I would probably go with more explicit and custom code for such a structure (as opposed to using "with" construct), but as a basic proof - simply changing the order of the fields makes the example work:

https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AbEAzAzmgFxCgFcA7NAExAGoAfCABxjIAIBlAT1wJgFsAdACVyBAJZ8YAgJJleUJuxhQAbmLAxcAWABQugNoAedgVJgCAbg6mS5gDIBDThBIEAFI+euA0mLKUBAFEERgx1MQIASisAOQhAgEcSBzCCTliIAGEIPkYHKDFcCDIAPgBdXVY05lYAEQkWXDFigCZa1gBeVkrWXr7+1gBvAZGBntHRowAxMRgMSgB5bDwYAgBBAlMxYFcYNwAGSPLxiYGAdTFKAgALVhBWbAwIBwIAZhaT077p2fmllfWmwKO14bhQRwqOi+IwAEjAxABza4EO4PJ4vd6faE/OaLZa4VYbLYgvaHY5Q6H9ACy/GgnFRfgIqCxIwAviyBgB3CLXDn9HgvdSsSR8YDKVgALWUEE6Q1YNL4dNl+zsVguV1uXX22CscMRyOVOtY7IpXz5fV0GFWrAQCFlw3qkjITVatQEUoUrG5N3ltKg9K6AEYWq8UEgAKwAdgAHABOQP7QN2Y26VN6HSMApybBsABEAFI1rmbQggA===

**Comment by T-Gro** [comment]
But this case kind of asks for manually written API (by you) - even if the example by me works, the constellation of fields overwriting each other is confusing for the API user.
(e.g. the result of doing "with" and overwriting both Memory and some of the values at the same time...)

**Comment by Xyncgas** [comment]
> But this case kind of asks for manually written API (by you) - even if the example by me works, the constellation of fields overwriting each other is confusing for the API user. (e.g. the result of doing "with" and overwriting both Memory and some of the values at the same time...)

Yeah and also if I am now trying to change width / height in your example, they get overwritten by Memory. I see the alternative data structure being an alternative way to go for now, but it's worth mention that as discussed by @vzarytovskii this doesn't happen in C# for some reason (you can use explicit struct with record like semantic normally)

I realized that there is a template for issues, I didn't see it because I pressed the button that transform the discussion to an issue, but I am happy to write it according to the template
